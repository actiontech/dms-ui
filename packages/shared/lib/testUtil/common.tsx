/* eslint-disable no-console */
export const ignoreAntdFormValidateWarning = () => {
  const warning = console.warn;

  beforeAll(() => {
    console.warn = (...arg) => {
      if (typeof arg[0] === 'string' && arg[0].includes('async-validator:')) {
        return;
      }
      warning(...arg);
    };
  });

  afterAll(() => {
    console.warn = warning;
  });
};

export const ignoreAntdUseFormNotConnectedError = () => {
  const error = console.error;

  beforeAll(() => {
    console.error = (...arg) => {
      if (
        typeof arg[0] === 'string' &&
        arg[0].includes(
          'Instance created by `useForm` is not connected to any Form element. Forget to pass `form` prop?'
        )
      ) {
        return;
      }
      error(...arg);
    };
  });

  afterAll(() => {
    console.error = error;
  });
};

export const ignoreComponentUncontrolled = () => {
  const error = console.error;

  beforeAll(() => {
    console.error = (...arg) => {
      if (
        typeof arg[0] === 'string' &&
        arg[0].includes(
          'A component is changing a controlled input to be uncontrolled.'
        )
      ) {
        return;
      }
      error(...arg);
    };
  });

  afterAll(() => {
    console.error = error;
  });
};

// 当为原生的 html 添加大驼峰、小驼峰之类的自定义属性，会出现报错
export const ignoreComponentCustomAttr = () => {
  const error = console.error;
  beforeAll(() => {
    console.error = (...arg) => {
      if (
        typeof arg[0] === 'string' &&
        arg[0].includes(
          'prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase'
        )
      ) {
        return;
      }
      error(...arg);
    };
  });

  afterAll(() => {
    console.error = error;
  });
};

/**
 * @description: 在选择 date-range-pick 组件的结束日期 input 元素时，实际选择到了，报此 warning
 */
export const ignoreComponentWarnExpDatePickChooseElement = () => {
  const warn = console.warn;
  beforeAll(() => {
    console.warn = (...arg) => {
      if (
        typeof arg[0] === 'string' &&
        arg[0].includes(
          'Not match any format. Please help to fire a issue about this.'
        )
      ) {
        return;
      }
      warn(...arg);
    };
  });

  afterAll(() => {
    console.warn = warn;
  });
};

export const ignoreComponentAutoCreatedListNoKey = () => {
  const error = console.error;

  beforeAll(() => {
    console.error = (...arg) => {
      if (
        typeof arg[0] === 'string' &&
        arg[0].includes('Each child in a list should have a unique "key" prop')
      ) {
        return;
      }
      error(...arg);
    };
  });

  afterAll(() => {
    console.error = error;
  });
};
