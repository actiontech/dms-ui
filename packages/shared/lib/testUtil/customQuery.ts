export {
  getBySelector,
  queryBySelector,
  getAllBySelector,
  getSelectContentByFormLabel,
  getSelectValueByFormLabel,
  selectOptionByIndex,
  selectCustomOptionByClassName,
  getValueByInput,
  sleep,
  getHrefByText,
  getAllHrefByText
} from '@actiontech/dms-kit/src/testUtil/customQuery';
