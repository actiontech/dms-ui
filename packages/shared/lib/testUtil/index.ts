export * from './superRender';
export * from './mockHook';
export * from './mockApi';
export * from './customQuery';
export * from './mockRedux';
export * from './common';
export * from './LocationDisplay';
