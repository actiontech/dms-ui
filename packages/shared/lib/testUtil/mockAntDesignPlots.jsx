import { cloneDeep, isObject, isBoolean } from 'lodash';

// 不适合作为 html 属性元素出现的 key 的数据集合
const customDomKeyData = ['label', 'legend'];

const MockPlots = (props) => {
  const { onReady, ref, ...othersParams } = cloneDeep(props);

  Object.keys(othersParams).forEach((key) => {
    if (customDomKeyData.includes(key)) {
      delete othersParams[key];
      return;
    }
    if (isObject(othersParams[key]) || isBoolean(othersParams[key])) {
      othersParams[key] = JSON.stringify(othersParams[key]);
    }
  });

  return <div {...othersParams} />;
};

const Line = MockPlots;
const Pie = MockPlots;
const RadialBar = MockPlots;
const Gauge = MockPlots;
const Column = MockPlots;
const Treemap = MockPlots;
const Area = MockPlots;
const Bar = MockPlots;
const RingProgress = MockPlots;
const mockRegisterShape = jest.fn();
const G2 = {
  registerShape: jest.fn()
};

export {
  Line,
  Pie,
  RadialBar,
  Gauge,
  Column,
  G2,
  mockRegisterShape,
  Treemap,
  Area,
  Bar,
  RingProgress
};

// eslint-disable-next-line import/no-anonymous-default-export
export default {
  Line,
  Pie,
  RadialBar,
  Gauge,
  Column,
  G2
};
