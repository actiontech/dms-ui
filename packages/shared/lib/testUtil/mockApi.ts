import { setupServer } from 'msw/node';
import { RestContext, RestHandler, MockedRequest, DefaultBodyType } from 'msw';
import MockBaseApi from './baseApi';
import { AxiosResponse } from 'axios';

class TestMockApi {
  private server!: ReturnType<typeof setupServer>;

  constructor() {
    this.initServer();
  }

  public getServer() {
    return this.server;
  }

  public initServer() {
    this.server = setupServer(...MockBaseApi.getHandler());
  }
}

export interface MockApi {
  getHandler(): RestHandler<MockedRequest<DefaultBodyType>>[];
}

export interface MockSpyApy {
  mockAllApi(): void;
}

export const createSuccessResponse = (ctx: RestContext, data?: any) => {
  return [
    ctx.delay(200),
    ctx.status(200),
    ctx.json({
      code: 0,
      msg: 'ok',
      payload: data
    })
  ];
};

export const createSpySuccessResponse = (
  data: any,
  { status = 200, headers = {}, config = {}, statusText = '' } = {}
) => {
  return new Promise<AxiosResponse<any>>((res) => {
    setTimeout(() => {
      res({
        status,
        headers,
        config,
        statusText,
        data: {
          code: 0,
          message: 'ok',
          ...data // ??？ 之前是 payload
        }
      });
    }, 3000);
  });
};

export const createSpyFailResponse = (
  data: any,
  { status = 200, headers = {}, config = {}, statusText = '' } = {}
) => {
  return new Promise<AxiosResponse<any>>((res) => {
    setTimeout(() => {
      res({
        status,
        headers,
        config,
        statusText,
        data: {
          code: 100,
          msg: 'error',
          payload: data
        }
      });
    }, 3000);
  });
};

export const createSpyErrorResponse = (
  data: any,
  { status = 200, headers = {}, config = {}, statusText = '' } = {}
) => {
  return new Promise<AxiosResponse<any>>((_, rej) => {
    setTimeout(() => {
      rej({
        status,
        headers,
        config,
        statusText,
        data: {
          code: 100,
          msg: 'error',
          payload: data
        }
      });
    }, 3000);
  });
};
export type MockServer = ReturnType<typeof setupServer>;

export default new TestMockApi();
