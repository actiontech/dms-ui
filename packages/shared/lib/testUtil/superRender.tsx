import { render, renderHook as renderHookReact } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MutableSnapshot, RecoilRoot } from 'recoil';
import { MemoryRouter, MemoryRouterProps } from 'react-router-dom';
import { ConfigProvider, theme as antdTheme } from 'antd';
import zhCN from 'antd/locale/zh_CN';
import { storeFactory } from './mockRedux';
import { ThemeProvider } from '@mui/system';
import { StyledEngineProvider } from '@mui/material/styles';
import lightTheme from '../theme/light';

export type RenderParams = Parameters<typeof render>;

export type RecoilRootPropsCustom = {
  initializeState?: (mutableSnapshot: MutableSnapshot) => void;
  override?: boolean;
};

export interface WrapperOptions {
  routerProps?: MemoryRouterProps;
  initStore?: any;
  recoilRootProps?: RecoilRootPropsCustom;
  theme?: any;
}

export const createTestWrapper = (options?: WrapperOptions) => {
  const {
    routerProps,
    initStore,
    recoilRootProps,
    theme = lightTheme
  } = options || {};

  return ({ children }: { children: React.ReactNode }) => {
    const themeComponent = (
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    );

    return (
      <ConfigProvider
        locale={zhCN}
        theme={{ algorithm: antdTheme.defaultAlgorithm, hashed: false }}
      >
        <RecoilRoot {...recoilRootProps}>
          <Provider store={storeFactory(initStore)}>
            <MemoryRouter {...routerProps}>
              <StyledEngineProvider injectFirst>
                {themeComponent}
              </StyledEngineProvider>
            </MemoryRouter>
          </Provider>
        </RecoilRoot>
      </ConfigProvider>
    );
  };
};

export const superRender = (
  ...[ui, option, options]: [...RenderParams, WrapperOptions?]
) => {
  const wrapper = createTestWrapper(options);

  return render(ui, {
    wrapper,
    ...option
  });
};

export const superRenderHook = <TProps, TResult>(
  hook: (props: TProps) => TResult,
  renderHookOptions?: Omit<
    Parameters<typeof renderHookReact<TProps, TResult>>[1],
    'wrapper'
  >,
  wrapperOptions?: WrapperOptions
) => {
  const wrapper = createTestWrapper(wrapperOptions);

  return renderHookReact(hook, {
    ...renderHookOptions,
    wrapper
  });
};
