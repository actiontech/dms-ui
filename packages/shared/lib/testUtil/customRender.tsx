import { render, renderHook } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MutableSnapshot, RecoilRoot } from 'recoil';
import {
  BrowserRouter,
  Router,
  RouterProps,
  MemoryRouter,
  MemoryRouterProps
} from 'react-router-dom';
import { ConfigProvider, theme as antdTheme } from 'antd';
import zhCN from 'antd/locale/zh_CN';
import { Dictionary, IStore } from '../types/common.type';
import { storeFactory } from './mockRedux';
import { StyledEngineProvider, ThemeProvider } from '@mui/system';
import { mount } from 'enzyme';

import lightTheme from '../theme/light';

type MountParams = Parameters<typeof mount>;
type RenderParams = Parameters<typeof render>;
type RecoilRootPropsCustom = {
  initializeState?: (mutableSnapshot: MutableSnapshot) => void;
  override?: boolean;
};

export const renderWithRouter = (...[ui, option]: [...RenderParams]) => {
  return render(<BrowserRouter>{ui}</BrowserRouter>, option);
};

export const renderWithServerRouter = (
  ...[ui, option, props]: [...RenderParams, RouterProps]
) => {
  return render(<Router {...props}>{ui}</Router>, option);
};

export const renderWithMemoryRouter = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(<MemoryRouter {...props}>{ui}</MemoryRouter>, option);
};

export const renderWithRedux = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(
    <Provider store={storeFactory(initStore)}>{ui}</Provider>,
    option
  );
};

export const renderWithReduxAndTheme = (
  ...[ui, option, initStore]: [...RenderParams, Dictionary?]
) => {
  return render(ui, {
    wrapper: ({ children }) => (
      <ThemeProvider theme={lightTheme}>
        <Provider store={storeFactory(initStore)}>{children}</Provider>
      </ThemeProvider>
    ),
    ...option
  });
};

export const renderHooksWithRedux = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: IStore
) => {
  return renderHook(hooks, {
    wrapper: ({ children }) => (
      <Provider store={storeFactory(storeState)}>{children}</Provider>
    )
  });
};

export const renderHooksWithReduxAndRouter = <TProps, TResult>(
  hooks: (props: TProps) => TResult,
  storeState: IStore
) => {
  return renderHook(hooks, {
    wrapper: ({ children }) => (
      <Provider store={storeFactory(storeState)}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    )
  });
};

export const renderHooksWithTheme = <TProps, TResult>(
  hooks: (props: TProps) => TResult
) => {
  return renderHook(hooks, {
    wrapper: ({ children }) => (
      <ConfigProvider
        locale={zhCN}
        theme={{ algorithm: antdTheme.defaultAlgorithm }}
      >
        <StyledEngineProvider injectFirst>
          <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
        </StyledEngineProvider>
      </ConfigProvider>
    )
  });
};

export const renderWithTheme = (...[ui, option]: [...RenderParams]) => {
  return render(<ThemeProvider theme={lightTheme}>{ui}</ThemeProvider>, option);
};

export const renderWithThemeAndRouter = (
  ...[ui, option, props]: [...RenderParams, MemoryRouterProps?]
) => {
  return render(
    <MemoryRouter {...props}>
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={lightTheme}>{ui}</ThemeProvider>
      </StyledEngineProvider>
    </MemoryRouter>,
    option
  );
};

export const superRender = (
  ...[ui, option, otherProps]: [
    ...RenderParams,
    {
      routerProps?: MemoryRouterProps;
      initStore?: any;
      recoilRootProps?: RecoilRootPropsCustom;
    }?
  ]
) => {
  const renderReturn = render(ui, {
    wrapper: ({ children }) => {
      return (
        <RecoilRoot {...otherProps?.recoilRootProps}>
          <Provider store={storeFactory(otherProps?.initStore)}>
            <MemoryRouter {...otherProps?.routerProps}>
              <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
            </MemoryRouter>
          </Provider>
        </RecoilRoot>
      );
    },
    ...option
  });
  return renderReturn;
};

export const mountWithTheme = (...[ui, option]: [...MountParams]) => {
  return mount(ui, {
    ...option,
    wrappingComponent: ThemeProvider,
    wrappingComponentProps: {
      theme: lightTheme
    }
  });
};
