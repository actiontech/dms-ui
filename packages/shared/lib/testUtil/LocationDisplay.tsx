import { RenderResult } from '@testing-library/react';
import { useLocation } from 'react-router-dom';
import { renderWithRouter } from './customRender';

export const renderLocationDisplay = (): [
  () => RenderResult,
  () => JSX.Element
] => {
  const LocationComponent = () => {
    const location = useLocation();

    return <div data-testid="location-display">{location.pathname}</div>;
  };

  return [() => renderWithRouter(<LocationComponent />), LocationComponent];
};
