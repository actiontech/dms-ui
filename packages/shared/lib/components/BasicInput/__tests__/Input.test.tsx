import { fireEvent, screen } from '@testing-library/react';
import { superRender } from '../../../testUtil/superRender';
import { BasicInput } from '..';
import { CloseOutlined } from '@actiontech/icons';

jest.mock('@actiontech/icons', () => ({
  CloseOutlined: jest.fn().mockImplementation(() => <span />)
}));

describe('lib/BasicInput/Input', () => {
  const customRender = (params = {}) => {
    return superRender(<BasicInput {...params} />);
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = customRender();
    expect(container).toMatchSnapshot();
    expect(screen.getByPlaceholderText('请输入{{name}}')).toBeInTheDocument();
  });

  it('should render basic input with default props', () => {
    const { container } = customRender();
    expect(container).toMatchSnapshot();
    expect(screen.getByPlaceholderText('请输入{{name}}')).toBeInTheDocument();
  });

  it('should render input with custom className', () => {
    const { container } = customRender({
      className: 'custom-input'
    });
    expect(container.querySelector('.custom-input')).toBeInTheDocument();
    expect(container.querySelector('.basic-input-wrapper')).toBeInTheDocument();
  });

  it('should render input with different sizes', () => {
    const { rerender } = customRender({ size: 'large' });
    expect(screen.getByRole('textbox')).toHaveClass('ant-input-lg');

    rerender(<BasicInput size="small" />);
    expect(screen.getByRole('textbox')).toHaveClass('ant-input-sm');

    rerender(<BasicInput size="middle" />);
    expect(screen.getByRole('textbox')).not.toHaveClass(
      'ant-input-lg',
      'ant-input-sm'
    );
  });

  it('should render input with clear icon when allowClear is true', () => {
    const { container } = customRender({
      allowClear: true,
      value: 'test'
    });
    expect(CloseOutlined).toHaveBeenCalledWith(
      expect.objectContaining({
        width: 14,
        height: 14,
        fill: 'currentColor'
      }),
      expect.anything()
    );
    expect(
      container.querySelector('.ant-input-clear-icon')
    ).toBeInTheDocument();
  });

  it('should call onChange when input value changes', () => {
    const onChange = jest.fn();
    customRender({ onChange });
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: 'test' } });
    expect(onChange).toHaveBeenCalled();
  });

  it('should render disabled input', () => {
    customRender({ disabled: true });
    expect(screen.getByRole('textbox')).toBeDisabled();
  });

  it('should render input with error status', () => {
    const { container } = customRender({ status: 'error' });
    expect(
      container.querySelector('.ant-input-status-error')
    ).toBeInTheDocument();
  });
});
