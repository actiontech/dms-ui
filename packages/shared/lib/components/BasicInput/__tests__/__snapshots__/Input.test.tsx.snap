// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lib/BasicInput/Input should match snapshot 1`] = `
<div>
  <input
    class="ant-input basic-input-wrapper css-1pd1cd5"
    placeholder="请输入{{name}}"
    type="text"
    value=""
  />
</div>
`;

exports[`lib/BasicInput/Input should render basic input with default props 1`] = `
<div>
  <input
    class="ant-input basic-input-wrapper css-1pd1cd5"
    placeholder="请输入{{name}}"
    type="text"
    value=""
  />
</div>
`;
